plugins/modules/cloud/misc/terraform.py

```diff
@@ -449,12 +474,67 @@ def main():
     if state == 'present' and module.params.get('parallelism') is not None:
         command.append('-parallelism=%d' % module.params.get('parallelism'))
 
-    variables_args = []
-    for k, v in variables.items():
-        variables_args.extend([
-            '-var',
-            '{0}={1}'.format(k, v)
-        ])
+    def process_args(variables, top):
+        variables_args = []
+        lowlevel_out = []
+        for k, v in variables.items():
+            if top:
+                if (isinstance(v, dict)):
+                    variables_args.extend([
+                        "-var",
+                        k + '={' + process_args(v, False) + '}'
+                    ])
+                if (isinstance(v, list)):
+                    l_out = []
+                    for item in v:
+                        l_out.append("{" + process_args(item, False) + "}")
+                    variables_args.extend([
+                        "-var",
+                        k + '=[' + ",".join(l_out) + ']'
+                    ])
+                if ((isinstance(v, int) or isinstance(v, float)) and not isinstance(v, bool)):
+                    variables_args.extend([
+                        "-var",
+                        '{0}={1}'.format(k, v)
+                    ])
+                if (isinstance(v, str)):
+                    variables_args.extend([
+                        "-var",
+                        '{0}={1}'.format(k, v)
+                    ])
+                if (isinstance(v, bool)):
+                    if v:
+                        variables_args.extend([
+                            "-var",
+                            '{0}=true'.format(k)
+                        ])
+                    else:
+                        variables_args.extend([
+                            "-var",
+                            '{0}=false'.format(k)
+                        ])
+
+            else:
+                if (isinstance(v, dict)):
+                    lowlevel_out.append(k + '={' + process_args(v, False) + '}')
+                if (isinstance(v, list)):
+                    lowlevel_out.append(k + '=[' + process_args(v, False) + ']')
+                if ((isinstance(v, int) or isinstance(v, float)) and not isinstance(v, bool)):
+                    lowlevel_out.append('{0}={1}'.format(k, v))
+                if (isinstance(v, str)):
+                    lowlevel_out.append('{0}="{1}"'.format(k, v))
```

Comment:
What happens if a string contains `"` or some other special symbol? You might want to use `json.dumps()` here instead of manually prepending and appending `"`.

Reply:
I hear what you're saying. I've not tested this with a special character in the string (such as '"'), which can break python level, but at the same time it can break Terraform as well.  To make sure the module retains its sanity:

lowlevel_out.append('{0}={1}'.format(k, json.dumps(v)))

This can work for `integer_types` as well as for boolean and strings.

Reply:
Handled within commit 76e71c772df7b6e31541826097260cc1e0cf84de